### DISCUSSION TITLE
How to use custom struct with raw sql query

### DISCUSSION THREAD

**Discussion by bcba25** [diesel-rs/diesel#2692] [Q&A]
Hello everyone. I got a question on raw queries. Haven’t found much on diesel’s documentation.

```
#[derive(Debug, QueryableByName)]
#[table_name = "products"]
struct Product {
    pub id: Uuid,
    pub user_id: Uuid,
    pub name: String,
    pub location: Option<GeogPoint>,
    pub distance: f64, <-- ERROR: cannot find type `distance` in module `products`, not found in `products`
}

const QUERY: &str = r#"SELECT "id", "user_id", "name", "location", ST_Distance(...) as "distance" FROM products WHERE ST_Distance(...) <= 1000 ORDER BY "distance" ASC"#; 
let products = diesel::sql_query(QUERY).load::<Product>(_connection);
```

I’m wondering how can I use my custom structs with diesel and raw queries. If I delete distance field from struct, it will compile and execute the query but I need to show distance as well, which does not exists on table and formed with a query itself. Could you suggest on this please

**Comment by weiznich** [comment]
I've moved this to the discussion part of the repository as this is not a bug report, but a question.
Otherwise [see the documentation](https://docs.diesel.rs/1.4.x/diesel/deserialize/trait.QueryableByName.html#deriving).

**Reply by bcba25** [reply]
@weiznich thanks and sorry for creating as a bug. Could you provide a simple example on how to do that ? Or probably a fix for my example. I believe it should be relatively simple but I didn't find a way to do that

**Reply by weiznich** [reply]
To cite from the documentation:
>To derive this trait, Diesel needs to know the SQL type of each field. You can do this by either annotating your struct with `#[table_name = "some_table"]` (in which case the SQL type will be `diesel::dsl::SqlTypeOf<table_name::column_name>`), **or by annotating each field with `#[sql_type = "SomeType"]`**.

(Bold by me)

That means you just need to annotate your distance field with some compatible sql type via `#[sql_type = "…"]`. I cannot give more details as it is unclear from your question what the `ST_Distance(…)` function actually returns.
